### PR TITLE
Systemd userdb

### LINKED ISSUE #246: Add support for userdb
Hello,

With the recent versions of systemd, there is a new `userdb` component added.

`libnss-systemd` is now trying to connect to a socket located in `/run/systemd/userdb/` that meas that any domain (including a user one) that should resolve user/group id might try to connect to it.

There is also an optional daemon running

Fedora policy already has support for this that adds custom types.

### PR DESCRIPTION
**Pull request by bauen1**
This only addresses the scenario where `userdb` is disabeled, e.g. debian.
Partially Fixes #246 

`init_unix_stream_socket_connectto` is used because `init_stream_connect` was already taken and grants additional (unwanted) access that is not necessary.

### CODE REVIEW

File: policy/modules/system/systemd.if [hunk 1]
```diff
@@ -426,6 +426,63 @@ interface(`systemd_signull_logind',`
...
+#
+interface(`systemd_stream_connect_userdb', `
+	gen_require(`
+		type systemd_userdb_runtime_t;
+	')
+
+	init_search_runtime($1)
+	allow $1 systemd_userdb_runtime_t:dir list_dir_perms;
+	allow $1 systemd_userdb_runtime_t:sock_file write_sock_file_perms;
+	init_unix_stream_socket_connectto($1)
```
Comment (pebenito) [inline policy/modules/system/systemd.if:483]: This would otherwise connect to `systemd-homed` if that exists?

**Comment by doverride** [comment]
Chris PeBenito <notifications@github.com> writes:

> @pebenito commented on this pull request.
>
> -------------------------------------------------------------------------------------------------------------------------------------
>
> In policy/modules/system/systemd.if:
>
>> +## </summary>
> +## <param name="domain">
> +##	<summary>
> +##	Domain allowed access.
> +##	</summary>
> +## </param>
> +#
> +interface(`systemd_stream_connect_userdb', `
> +	gen_require(`
> +		type systemd_userdb_runtime_t;
> +	')
> +
> +	init_search_runtime($1)
> +	allow $1 systemd_userdb_runtime_t:dir list_dir_perms;
> +	allow $1 systemd_userdb_runtime_t:sock_file write_sock_file_perms;
> +	init_unix_stream_socket_connectto($1)
>
> This would otherwise connect to systemd-homed if that exists?

https://systemd.io/USER_GROUP_API/

/run/systemd/userdb (systemd_user_db_runtime_t) is a location where
various sockets are stored.

In the bare essenence this location is used to store the
io.systemd.DynamicUser socket which is used to connect to systemd (pid1)
This socket is more recently used to resolve dynamic users (nss
password/group) on the system
(in the past dbus was used for this: refpolicy has the systemd_use_nss()
for this deprecated way to resolve dynamic users)

In systemd v246 io.systemd.Machine was added which is used by the nss
mymachines module to resolve dynamic uids associated with
container. This socket is used to connect to systemd-machined (optional if its installed)

The io.systemd.Multiplexer socket in there is used to connect to
systemd-userdbd (optional if it exists) This is mostly used by PAM login
programs, and systemd-userdbctl

The io.systemd.Homed socket is used to connect to systemd-homed
(optional if its installed) this is also used by nss password/group to
resolve uids managed with systemd-userdbd

In systemd v247 there will likely be a io.systemd.Resolve added that can
be used to connect to systemd-resolve for (nss hosts: resolve module) to
resolve host names

So various sockets in /run/systemd/userdb all used to connect to
different targets of which almost everyone is optional one way or
another (except for the io.systemd.DynamicUser)

>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.
>

-- 
gpg --locate-keys dominick.grift@defensec.nl
Key fingerprint = FCD2 3660 5D6B 9D27 7FC6  E0FF DA7E 521F 10F6 4098
https://sks-keyservers.net/pks/lookup?op=get&search=0xDA7E521F10F64098
Dominick Grift

**Comment by pebenito** [comment]
In that case, it seems like the directory should keep the more general `init_runtime_t` and there should be named filetrans for each of the `sock_file`s.

**Comment by doverride** [comment]
Chris PeBenito <notifications@github.com> writes:

> In that case, it seems like the directory should keep the more general init_runtime_t and there should be named filetrans for each of
> the sock_files.

That is a possibility. I decided to not go that route in my personal
policy as there is the connectto check on the process unix_stream_socket
that can be used to differentiate. Figured creating private types for
each socket would be expensive.

Not to mention that you might end up with long type names for the
sockets to be able to differentiate. Take for example systemd sockets. there is already the notify
socket for Type=notify services, and the private socket for system
manager talk, and then now also the the dynamicuser socket for resolving
names.

>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or unsubscribe.
>

-- 
gpg --locate-keys dominick.grift@defensec.nl
Key fingerprint = FCD2 3660 5D6B 9D27 7FC6  E0FF DA7E 521F 10F6 4098
https://sks-keyservers.net/pks/lookup?op=get&search=0xDA7E521F10F64098
Dominick Grift

**Comment by pebenito** [comment]
Thanks for the insight @doverride. I was starting to have the feeling that labeling the sockets differently might be problematic too.

**Comment by pebenito** [comment]
I'm not sure where we're at with this PR?

**Comment by bauen1** [comment]
> I'm not sure where we're at with this PR?

I've stopped using refpolicy in favour of my own hand written policy, so I don't really know, but apart from that it should still be good to merge after a quick test.

**Comment by minimaxwell** [comment]
Hi all, I've been working on using the refpolicy on embedded systems, generated with Buildroot, following the steps of [Antoine Ténart](https://github.com/atenart).

I've been using a branch with these patches for a while now, and it helps by easily allowing systemd-udevd and dbus-daemon to connect to userdb, as discussed [here](https://github.com/SELinuxProject/refpolicy/pull/307#discussion_r483456496)

So this PR definitely helps, and at least on my side, i've been able to successfully test it.

I hope this can help the PR getting closer to being merged.

Thanks,

Maxime

**Pull request merged**